Add SFTP guide for JetBrains IDEs (#15588)
gravitational/teleport: docs/pages/server-access/guides/jetbrains-sftp.mdx

---
title: JetBrains SFTP
description: How to use JetBrains IDE to browse remote host with Teleport
h1: SFTP with JetBrains IDE
---

> This guide explains how to use Teleport and JetBrains IDE to access files with SFTP.

Reviewer: ```suggestion
This guide explains how to use Teleport and a JetBrains IDE to access files with SFTP.
```

> This guide explains how to use Teleport and JetBrains IDE to access files with SFTP.

Reviewer: For users who discover this guide after already visiting the docs site, I think we should add another sentence to the first paragraph about why using Teleport with SFTP and a JetBrains IDE is useful.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

> - JetBrains IDE like PyCharm, IntelliJ, Goland etc. [Products](https://www.jetbrains.com/products/#type=ide)

Reviewer: ```suggestion
- JetBrains IDE like PyCharm, IntelliJ, Goland etc. See [Products](https://www.jetbrains.com/products/#type=ide) for a full list of JetBrains IDEs.
```
Otherwise it's not clear what this link points to.

- One or more Teleport Nodes with Server Access enabled. If you have not yet
done this, read the [Server Access Getting Started Guide](../getting-started.mdx) to learn how.

## Step 1/3. First-time setup

Configure your local SSH client to access Teleport Nodes, assigning the `--proxy` flag to the address of your Teleport Proxy Service (e.g., `mytenant.teleport.sh` for Teleport Cloud users).

```code
# log in to your proxy:
$ tsh login --proxy proxy.foo.example.com --user alice

# generate the OpenSSH config for the proxy:
$ tsh config --proxy proxy.foo.example.com
```

Append the resulting configuration snippet into your SSH config file located
in the path below:

<Tabs>
  <TabItem label="Linux/MacOS">
    `$HOME/.ssh/config`
  </TabItem>
  <TabItem label="Windows">
    `%UserProfile%\.ssh\config`

    <Admonition type="warning">
      If using PowerShell on Windows to write your SSH config, note that normal
      shell redirection may write the file with the incorrect encoding. To
      ensure it's written properly, try the following:

      ```code
      $ tsh.exe config | out-file .ssh\config -encoding utf8 -append
      ```
    </Admonition>
  </TabItem>
</Tabs>

> You should be able to connect to the desired node using following command, replacing `user` with the username you would like to assume on the node.

Reviewer: ```suggestion
You should be able to connect to the desired Node using following command, replacing `user` with the username you would like to assume on the Node:
```
Using "Node" so Teleport SSH Service instances are named consistently in the docs.

```code
$ ssh user@[node name].[cluster name]
```

<Details scopeOnly={true} scope={["cloud"]} title="Teleport Cloud">
  The SSH config you generated earlier instructs your SSH client to run `tsh proxy ssh` to access a Node in your Teleport cluster.
  However, running an `ssh` command against the Teleport Proxy Service at `yourtenant.teleport.sh` will result in an error.
</Details>

<Admonition type="note">
  Teleport's certificates expire fairly quickly, after which SSH
  attempts will fail with an error like the following:

  ```txt
  alice@proxy.foo.example.com: Permission denied (publickey).
  ERROR: exit status 255

  kex_exchange_identification: Connection closed by remote host
  ```

  When you see this error, re-run `tsh login` to refresh your local certificate.
</Admonition>

> ## Step 2/3. Configure JetBrains IDE

Reviewer: ```suggestion
## Step 2/3. Configure your JetBrains IDE
```

After opening your IDE go to `Tools` -> `Deployment` -> `Browse Remote Host`.

![Deployment](../../../img/jetbrains-sftp/deployment-main.png)

> Then click plus sign in the left top corner to add a new server.

Reviewer: ```suggestion
Then click the plus sign in the top-left corner to add a new server.
```

![Add server](../../../img/jetbrains-sftp/add-server.png)

Enter a name for your new server.

![New Deployment](../../../img/jetbrains-sftp/deployment-added.png)

> Click three dots next to `SSH configuration` as in the picture above.

Reviewer: ```suggestion
Click the three dots next to `SSH configuration` as in the picture above.
```

![SSH Configuration](../../../img/jetbrains-sftp/ssh-configurations.png)

> Create a new configuration by clicking plus sign on the left top and providing:

Reviewer: ```suggestion
Create a new configuration by clicking the plus sign on the top left and providing:
```

> Create a new configuration by clicking plus sign on the left top and providing:

Reviewer: I think the items in this list should use ` - ` and "is" consistently to demarcate fields from their descriptions. Personally, I think using "is" instead of ` - ` works a bit better here.

> * `Host` - name of the Teleport node.

Reviewer: ```suggestion
* `Host` - name of the Teleport Node.
```

> * `Port` - is the SSH node port (3022 by default).

Reviewer: ```suggestion
* `Port` - is the Teleport Node's SSH port (3022 by default).
```

* `Username` - is the system username.

As an `Authentication type` pick `OpenSSH config and authentication agent`.
Next, you can click `Test Connection`.

![Successfully Connected](../../../img/jetbrains-sftp/successfully-connected.png)

> ## Step 3/3. Browse remote host

Reviewer: ```suggestion
## Step 3/3. Browse a remote host
```

---
title: JetBrains SFTP

> description: How to use JetBrains IDE to browse remote host with Teleport

Reviewer: ```suggestion
description: How to use a JetBrains IDE to manipulate files on a remote host with Teleport
```

h1: SFTP with JetBrains IDE
---

> JetBrain's IDEs, like PyCharm, Goland, Intellij, allow browsing, copying, and editing files on a remote server

Reviewer: ```suggestion
JetBrain's IDEs, like PyCharm, GoLand, and IntelliJ, allow browsing, copying, and editing files on a remote server
```

> using SFTP protocol. The built-in integration will simplify the process if you are used to copying files back and forth

Reviewer: ```suggestion
using the SFTP protocol. The built-in integration will simplify the process if you are used to copying files back and forth
```

> using `scp` or other 3rd party clients. You can do the same things without leaving your IDE anymore.

Reviewer: I think we can condense the final two sentences of this paragraph, maybe something like:

```markdown
You can integrate Teleport Server Access with your 
IDE so you can copy files to and from a remote 
machine without using a third-party client.
```

This guide explains how to use Teleport and a JetBrains IDE to access files with SFTP.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

> - JetBrains IDE like PyCharm, IntelliJ, Goland etc. See [Products](https://www.jetbrains.com/products/#type=ide) for a full list of JetBrains IDEs.

Reviewer: ```suggestion
- JetBrains IDE like PyCharm, IntelliJ, GoLand etc. See [Products](https://www.jetbrains.com/products/#type=ide) for a full list of JetBrains IDEs.
```

- One or more Teleport Nodes with Server Access enabled. If you have not yet
done this, read the [Server Access Getting Started Guide](../getting-started.mdx) to learn how.

## Step 1/3. First-time setup

Configure your local SSH client to access Teleport Nodes, assigning the `--proxy` flag to the address of your Teleport Proxy Service (e.g., `mytenant.teleport.sh` for Teleport Cloud users).

```code
# log in to your proxy:
$ tsh login --proxy proxy.foo.example.com --user alice

# generate the OpenSSH config for the proxy:
$ tsh config --proxy proxy.foo.example.com
```

Append the resulting configuration snippet into your SSH config file located
in the path below:

<Tabs>
  <TabItem label="Linux/MacOS">
    `$HOME/.ssh/config`
  </TabItem>
  <TabItem label="Windows">
    `%UserProfile%\.ssh\config`

    <Admonition type="warning">
      If using PowerShell on Windows to write your SSH config, note that normal
      shell redirection may write the file with the incorrect encoding. To
      ensure it's written properly, try the following:

      ```code
      $ tsh.exe config | out-file .ssh\config -encoding utf8 -append
      ```
    </Admonition>
  </TabItem>
</Tabs>

> You should be able to connect to the desired Node using following command, replacing `user` with the username you would like to assume on the Node:

Reviewer: ```suggestion
You should be able to connect to the desired Node using the following command, replacing `user` with the username you would like to assume on the Node:
```

</Details>

<Admonition type="note">
  Teleport's certificates expire fairly quickly, after which SSH
  attempts will fail with an error like the following:

  ```txt
  alice@proxy.foo.example.com: Permission denied (publickey).
  ERROR: exit status 255

  kex_exchange_identification: Connection closed by remote host
  ```

  When you see this error, re-run `tsh login` to refresh your local certificate.
</Admonition>

## Step 2/3. Configure your JetBrains IDE

After opening your IDE go to `Tools` -> `Deployment` -> `Browse Remote Host`.

![Deployment](../../../img/jetbrains-sftp/deployment-main.png)

Then click the plus sign in the top-left corner to add a new server.

![Add server](../../../img/jetbrains-sftp/add-server.png)

Enter a name for your new server.

![New Deployment](../../../img/jetbrains-sftp/deployment-added.png)

Click the three dots next to `SSH configuration` as in the picture above.

![SSH Configuration](../../../img/jetbrains-sftp/ssh-configurations.png)

Create a new configuration by clicking the plus sign on the top left and providing:

- `Host` - name of the Teleport Node.

> - `Port` - is the Teleport Node's SSH port (3022 by default).

Reviewer: ```suggestion
- `Port` - the Teleport Node's SSH port (3022 by default).
```

> - `Username` - is the system username.

Reviewer: ```suggestion
- `Username` - the system username.
```

As an `Authentication type` pick `OpenSSH config and authentication agent`.
Next, you can click `Test Connection`.

![Successfully Connected](../../../img/jetbrains-sftp/successfully-connected.png)

## Step 3/3. Browse a remote host

After closing the SSH configuration window, you should see `Remote Host` menu in your IDE.

![Browse window](../../../img/jetbrains-sftp/browse-window.png)

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```
